go/gilded-rose_test.go

```diff
@@ -1,15 +1,106 @@
+		{testname: "Aged Brie 0 40", in: []*Item{{name: "Aged Brie", sellIn: 0, quality: 40}}, want: []*Item{{name: "Aged Brie", sellIn: -1, quality: 42}}},
+		{testname: "Backstage passes to a TAFKAL80ETC concert 0 10", in: []*Item{{name: "Backstage passes to a TAFKAL80ETC concert", sellIn: 0, quality: 10}}, want: []*Item{{name: "Backstage passes to a TAFKAL80ETC concert", sellIn: -1, quality: 0}}},
+		{testname: "A 0 10", in: []*Item{{name: "A", sellIn: 0, quality: 10}}, want: []*Item{{name: "A", sellIn: -1, quality: 8}}},
+		{testname: "Conjured 0 10", in: []*Item{{name: "Conjured", sellIn: 0, quality: 10}}, want: []*Item{{name: "Conjured", sellIn: -1, quality: 6}}},
+		{testname: "Conjured 1 10", in: []*Item{{name: "Conjured", sellIn: 1, quality: 10}}, want: []*Item{{name: "Conjured", sellIn: 0, quality: 8}}},
+		{testname: "Conjured 0 10", in: []*Item{{name: "Conjured", sellIn: 0, quality: 2}}, want: []*Item{{name: "Conjured", sellIn: -1, quality: 0}}},
 	}
 
-	UpdateQuality(items)
+	for testInd, tt := range tests {
+		t.Run(tt.testname, func(t *testing.T) {
+			UpdateQuality(tt.in)
+			for i := range tt.in {
+				if !reflect.DeepEqual(tt.in[i], tt.want[i]) {
+					t.Errorf("test index:%v, got %v, want %v", testInd, tt.in[i], tt.want[i])
+				}
+			}
+		})
+	}
+}
+
+func TestAddQuality(t *testing.T) {
+	var tests = []struct {
+		testname string
+		in       *Item
+		change   int
+		want     error
+	}{
+		{testname: "A 1 1, addQuality -1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: -1, want: fmt.Errorf("addQuality: val is negative, %v", -1)},
+		{testname: "A 1 1, addQuality 1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: 1, want: nil},
+	}
+
+	for testInd, tt := range tests {
+		t.Run(tt.testname, func(t *testing.T) {
+			err := tt.in.addQuality(tt.change)
+			if !reflect.DeepEqual(err, tt.want) {
+				t.Errorf("test index:%v, got %v, want %v", testInd, err, tt.want)
+			}
+		})
+	}
+}
+
+func TestSubQuality(t *testing.T) {
+	var tests = []struct {
+		testname string
+		in       *Item
+		change   int
+		want     error
+	}{
+		{testname: "A 1 1, subQuality -1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: -1, want: fmt.Errorf("subQuality: val is negative, %v", -1)},
+		{testname: "A 1 1, subQuality 1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: 1, want: nil},
+	}
+
+	for testInd, tt := range tests {
+		t.Run(tt.testname, func(t *testing.T) {
+			err := tt.in.subQuality(tt.change)
+			if !reflect.DeepEqual(err, tt.want) {
+				t.Errorf("test index:%v, got %v, want %v", testInd, err, tt.want)
+			}
+		})
+	}
+}
+
+func TestSubSellin(t *testing.T) {
+	var tests = []struct {
+		testname string
+		in       *Item
+		change   int
+		want     error
+	}{
+		{testname: "A 1 1, subSellIn -1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: -1, want: fmt.Errorf("subSellIn: val is negative, %v", -1)},
+		{testname: "A 1 1, subSellIn 1", in: &Item{name: "A", sellIn: 1, quality: 1}, change: 1, want: nil},
```

Comment:
add tests cases for `change: 1` where `&Item{name: "A", sellIn: 1, quality: 0}` what would you expect the logic to perform?